tools/bpf2c/bpf_code_generator.h

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#pragma once
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+#include "ebpf.h"
+#include "ebpf_structs.h"
+#pragma warning(push)
+#pragma warning(disable : 4458) /* declaration of 'name' hides class member */
+#include "elfio/elfio.hpp"
+#pragma warning(pop)
+
+class bpf_code_generator
+{
+  public:
+    /**
+     * @brief Construct a new bpf code generator object.
+     *
+     * @param[in] path Path to the eBPF file to parse.
+     * @param[in] c_name C compatible name to export this as.
+     */
+    bpf_code_generator(const std::string& path, const std::string& c_name);
+
+    /**
+     * @brief Construct a new bpf code generator object from raw eBPF byte code.
+     *
+     * @param[in] c_name C compatible name to export this as.
+     * @param[in] instructions Set of eBPF instructions to use.
+     */
+    bpf_code_generator(const std::string& c_name, const std::vector<ebpf_inst>& instructions);
+
+    std::vector<std::string>
+    program_sections();
+
+    /**
+     * @brief Parse the eBPF file.
+     *
+     * @param[in] section_name Section in the ELF file to parse.
+     */
+    void
+    parse(const std::string& section_name);
+
+    /**
+     * @brief Generate C code from the parsed eBPF file.
+     *
+     */
+    void
+    generate();
+
+    /**
+     * @brief Emit the C code to a given output stream.
+     *
+     * @param[in] output Output stream to write code to.
+     */
+    void
+    emit_c_code(std::ostream& output);
+
+  private:
+    typedef struct _output_instruction
+    {
+        ebpf_inst instruction = {};
+        uint32_t instruction_offset;
+        bool jump_target = false;
+        std::string label;
+        std::vector<std::string> lines;
+        std::string relocation;
+    } output_instruction_t;
+
+    typedef struct _section
+    {
+        std::vector<output_instruction_t> output;
+        std::set<std::string> referenced_registers;
+    } section_t;
+
+    typedef struct _helper_function
+    {
+        int32_t id;
+        size_t index;
+    } helper_function_t;
+
+    typedef struct _map_entry
+    {
+        ebpf_map_definition_in_file_t definition;
+        size_t index;
+    } map_entry_t;
+
+    /**
+     * @brief Extract the eBPF byte code from the eBPF file.
+     *
+     */
+    void
+    extract_program(const std::string& section_name);
+
+    /**
+     * @brief Extract the helper function and map relocation data from the eBPF file.
+     *
+     */
+    void
+    extract_relocations_and_maps(const std::string& section_name);
+
+    // TODO - Once https://github.com/microsoft/ebpf-for-windows/issues/630 is
+    // resolved, gather and emit corresponding C code if available.
```

Comment:
I think the relationship should be that if there's a way to get the original source filename & line number embedded into the PE binary, then source level debugging could use the original source file instead of the generated source file, and hence be a better developer experience.

Reply:
I think we could do this with a [#pragma comment(user, "file#line")](https://docs.microsoft.com/en-us/cpp/preprocessor/comment-c-cpp?view=msvc-170). 

According to the docs, this will emit the text into the resulting executable image. Will investigate if there is an option to emit this to the PDB instead.

Reply:
Apparently there is also the [#line directive](https://docs.microsoft.com/en-us/cpp/preprocessor/hash-line-directive-c-cpp?view=msvc-170).

This seems like it's tailor-made for this exact scenario :-)

Reply:
It looks like the VS2019 debugger honors the #line directive, so once we figure out how to decode the dwarfv4 data in the ELF file we should be able to support source-level debugging of eBPF programs.